protocols/kad/src/kbucket.rs

```diff
@@ -161,254 +176,502 @@ where
+                    *first_connected_pos -= 1;
+                    table.nodes.insert(*first_connected_pos, pending.node);
+                } else {
+                    table.nodes.insert(*first_connected_pos - 1, pending.node);
+                }
+            } else {
+                self.tables[bucket_num].pending_node = Some(pending);
+            }
+        }
+
+        // Try to find the node in the bucket.
+        if let Some(pos) = self.tables[bucket_num].nodes.iter().position(|p| p.id == *peer_id) {
+            if pos >= self.tables[bucket_num].first_connected_pos {
+                Entry::InKbucketConnected(EntryInKbucketConn {
+                    parent: self,
+                    peer_id,
+                })
+
+            } else {
+                Entry::InKbucketDisconnected(EntryInKbucketDisc {
+                    parent: self,
+                    peer_id,
+                })
+            }
+
+        } else if self.tables[bucket_num].pending_node.as_ref().map(|p| p.node.id == *peer_id).unwrap_or(false) {
+            // Node is pending.
+            if self.tables[bucket_num].pending_node.as_ref().map(|p| p.connected).unwrap_or(false) {
+                Entry::InKbucketConnectedPending(EntryInKbucketConnPending {
+                    parent: self,
+                    peer_id,
+                })
+            } else {
+                Entry::InKbucketDisconnectedPending(EntryInKbucketDiscPending {
+                    parent: self,
+                    peer_id,
+                })
+            }
+
+        } else {
+            Entry::NotInKbucket(EntryNotInKbucket {
+                parent: self,
+                peer_id,
+            })
+        }
+    }
+
+    /// Returns an iterator to all the peer IDs in the bucket, without the pending nodes.
+    pub fn entries_not_pending(&self) -> impl Iterator<Item = (&TPeerId, &TVal)> {
+        self.tables
+            .iter()
+            .flat_map(|table| table.nodes.iter())
+            .map(|node| (&node.id, &node.value))
+    }
+
     /// Returns an iterator to all the buckets of this table.
     ///
     /// Ordered by proximity to the local node. Closest bucket (with max. one node in it) comes
     /// first.
-    #[inline]
     pub fn buckets(&mut self) -> BucketsIter<'_, TPeerId, TVal> {
         BucketsIter(self.tables.iter_mut(), self.unresponsive_timeout)
     }
 
-    /// Returns the ID of the local node.
-    #[inline]
-    pub fn my_id(&self) -> &TPeerId {
-        &self.my_id
+    /// Finds the nodes closest to `id`, ordered by distance.
+    ///
+    /// Pending nodes are ignored.
+    pub fn find_closest(&mut self, id: &impl KBucketsPeerId<TPeerId>) -> VecIntoIter<TPeerId> {
+        // TODO: optimize
+        let mut out = Vec::new();
+        for table in self.tables.iter_mut() {
+            for node in table.nodes.iter() {
+                out.push(node.id.clone());
+            }
+            // TODO: that's kind of wrong
```

Comment:
Could you be more specific?

Reply:
We need the same logic here as in the `entry` method. The code here works but doesn't have the same behaviour, which is a bit stupid.

In practice though, we need a way to report when a pending node replaces a valid node (there's another TODO for that), meaning that we need to rework this anyway.

Reply:
Maybe this should be mentioned in the comment?

Reply:
Done